[RDY] Use uv_getaddrinfo() for servers

mhinz opened the pull request:
This change implicitly adds IPv6 support.

If the address contains ":", we try to use a TCP socket instead of a Unix domain
socket. Everything in front of the last occurrence of ":" is the hostname and
everything after it the port.

If the hostname lookup fails, we fall back to using a Unix domain socket.

If the port is empty ("localhost:"), a random port will be assigned.

Examples:

```
  NVIM_LISTEN_ADDRESS=localhost:12345  -> TCP (IPv4 or IPv6), port: 12345
  NVIM_LISTEN_ADDRESS=localhost:       -> TCP (IPv4 or IPv6), port: random (> 1024)
  NVIM_LISTEN_ADDRESS=localhost        -> Unix domain socket "localhost" in current dir
```

mhinz committed:
Server: use uv_getaddrinfo() for $NVIM_LISTEN_ADDRESS

This change implicitly adds IPv6 support.

If the address contains ":", we try to use a TCP socket instead of a Unix domain
socket. Everything in front of the last occurrence of ":" is the hostname and
everything after it the port.

If the hostname lookup fails, we fall back to using a Unix domain socket.

If the port is empty ("localhost:"), a random port will be assigned.

Examples:

  NVIM_LISTEN_ADDRESS=localhost:12345 -> TCP (IPv4 or IPv6), port: 12345
  NVIM_LISTEN_ADDRESS=localhost:      -> TCP (IPv4 or IPv6), port: random (> 1024)
  NVIM_LISTEN_ADDRESS=localhost:0     -> TCP (IPv4 or IPv6), port: random (> 1024)
  NVIM_LISTEN_ADDRESS=localhost       -> Unix domain socket "localhost" in current dir

jamessan commented:
Should we be using `uv_getaddrinfo` instead, which would also support IPv6?

mhinz commented:
I look into it!

oni-link commented:
See function `socket_connect()` in #6594 (https://github.com/neovim/neovim/pull/6594/files#diff-6d5b2f9d654e4e24ddb9a960da02581eR182), it already uses `uv_getaddrinfo()` for address lookup.

mhinz commented:
@oni-link Thanks for the pointer!

Nevertheless, I think think this will take me more time than I'm willed to invest at the moment.

Does someone want to take over? Maybe @bfredl? Seems like a good addition for the already mentioned https://github.com/neovim/neovim/pull/6594.

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -90,14 +71,38 @@ int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
+    for (; ai; ai = ai->ai_next) {
+      memcpy(&watcher->uv.tcp.addr, ai->ai_addr, ai->ai_addrlen);
+      result = uv_tcp_bind(&watcher->uv.tcp.handle,
+                           (const struct sockaddr *)&watcher->uv.tcp.addr, 0);
```
`ai->ai_addr` can be used here.

mhinz replied:
Right, I guess it would be more explicit to use `watcher` since it's also used for all the following calls.. but using `ai->ai_addr` I can avoid the cast and it just looks better.

:+1:

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -90,14 +71,38 @@ int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
   if (watcher->stream->type == UV_TCP) {
-    result = uv_tcp_bind(&watcher->uv.tcp.handle,
-                         (const struct sockaddr *)&watcher->uv.tcp.addr, 0);
+    struct addrinfo *ai = watcher->uv.tcp.addrinfo;
```
If `ai` is `NULL` then `result` is not initialized.

mhinz replied:
Oops. I'll fix that.

mhinz replied:
Since libuv is expected to return negative integers, I now initialise it with `-EINVAL`.

ZyX-I replied:
@mhinz This should be UV_EINVAL without `-`, not `-EINVAL`. According to C99 only EDOM, EILSEQ and ERANGE must be defined, no EINVAL in this list.

ZyX-I replied:
Also on Windows UV_EINVAL and -EINVAL are going to be different AFAIK. And you return UV errors in other places. Also UV_EACCESS and not `-EACCESS` below, UV_ENOENT and not `-ENOENT`.

mhinz replied:
:+1:

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -90,14 +71,38 @@ int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
+                           &(int){sizeof(struct sockaddr_storage)});
+        uint16_t port = (ai->ai_family == AF_INET)
+          ? ((struct sockaddr_in  *)&watcher->uv.tcp.addr)->sin_port
+          : ((struct sockaddr_in6 *)&watcher->uv.tcp.addr)->sin6_port;
```
These numbers are encoded in network byte order, what we want is host byte order, so we need to call `ntohs()` on `sin_port/sin6_port`.

mhinz replied:
`ntohs()` is used when `port` is used. I wanted to avoid wrapping this already huge assignment and using an extra assignment à la `port = ntohs(port)`.

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -90,14 +71,38 @@ int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
+        // number, a free random port number will be assigned. sin_port will
+        // contain 0 in this case, unless uv_tcp_getsockname() is used first.
+        uv_tcp_getsockname(&watcher->uv.tcp.handle,
+                           (struct sockaddr *)&watcher->uv.tcp.addr,
```
Why not use a local `struct sockaddr_storage x` here? The content of `x` is not used in this call. Member `watcher->uv.tcp.addr` could be removed.

mhinz replied:
:+1:

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -81,23 +57,48 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+          : ((struct sockaddr_in6 *)&watcher->uv.tcp.addr)->sin6_port;
+        // v:servername uses the string from watcher->addr
+        snprintf(watcher->addr, sizeof(watcher->addr), "%s:%" PRIu16,
+                 watcher->addr, ntohs(port));
```
Hostname is not showing for me. Could be that we reading and writing from the same buffer.

mhinz replied:
Hmm. Okay, I'll make a copy of `watcher->addr` first.

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -81,23 +57,48 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+    struct addrinfo *ai = watcher->uv.tcp.addrinfo;
+
+    for (; ai; ai = ai->ai_next) {
+      memcpy(&watcher->uv.tcp.addr, ai->ai_addr, ai->ai_addrlen);
```
This could be removed if we use a local variable, see below.

mhinz replied:
:+1:

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -81,23 +57,49 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+    struct addrinfo *ai = watcher->uv.tcp.addrinfo;
+
+    for (; ai; ai = ai->ai_next) {
+      struct sockaddr_storage sas = *(struct sockaddr_storage *)ai->ai_addr;
```
This should be moved into the `if (result == 0) {...}` block without initialization (`uv_tcp_getsockname()` will fill it). Dereferencing `ai->ai_addr` can go wrong if the structure is smaller than `struct sockaddr_storage`.

mhinz replied:
> Dereferencing ai->ai_addr can go wrong if the structure is smaller than struct sockaddr_storage.

How would I work around that? A `union` with `sockaddr_in` and `sockaddr_in6` and switching on `ai_family`?

oni-link replied:
Just put `struct sockaddr_storage sas;` before `uv_tcp_getsockname()`. We get the socket name from the socket descriptor, not from the content of `sas`.

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -81,23 +57,49 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+        // v:servername uses the string from watcher->addr
+        char addr[sizeof(watcher->addr)];
+        strcpy(addr, watcher->addr);
+        snprintf(watcher->addr, sizeof(addr), "%s:%" PRIu16, addr, ntohs(port));
```
Could also be done with a `strlen()`:
```c
size_t len = strlen(watcher->addr);
snprintf(watcher->addr +len, sizeof(watcher->addr)-len, ":%" PRIu16, ntohs(port));
```

mhinz replied:
:+1:

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -21,58 +21,34 @@
-    const char *endpoint, void *data)
+void socket_watcher_init(Loop *loop, SocketWatcher *watcher, const char *endpoint,
+                         void *data)
   FUNC_ATTR_NONNULL_ARG(1) FUNC_ATTR_NONNULL_ARG(2) FUNC_ATTR_NONNULL_ARG(3)
```
We can use `FUNC_ATTR_NONNULL_ARG(1,2,3)` here.

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -21,58 +21,34 @@
-void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
-    const char *endpoint, void *data)
+void socket_watcher_init(Loop *loop, SocketWatcher *watcher, const char *endpoint,
+                         void *data)
```
Are we missing a `watcher->data = data`?

mhinz replied:
Can we just remove `data` from the arguments? It's only called with `NULL` in `server_start()`. Maybe the original author planned ahead when he wrote that.

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -21,58 +21,34 @@
-    if (uv_ip4_addr(ip, port, &watcher->uv.tcp.addr)) {
-      // Invalid address, treat as named pipe or unix socket
-      tcp = false;
+  char *addr = xstrdup(endpoint);
```
We could copy endpoint directly into `watcher->addr` and work with that string:
```diff
diff --git a/src/nvim/event/socket.c b/src/nvim/event/socket.c
index 48d4c3716..c254b8077 100644
--- a/src/nvim/event/socket.c
+++ b/src/nvim/event/socket.c
@@ -25,12 +25,12 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher, const char *endpoin
                          void *data)
   FUNC_ATTR_NONNULL_ARG(1) FUNC_ATTR_NONNULL_ARG(2) FUNC_ATTR_NONNULL_ARG(3)
 {
-  char *addr = xstrdup(endpoint);
+  xstrlcpy(watcher->addr, endpoint, sizeof(watcher->addr));
+  char *addr = watcher->addr;
   char *host_end = strrchr(addr, ':');
 
   if (host_end && addr != host_end) {
     *host_end = '\0';
-    xstrlcpy(watcher->addr, addr, sizeof(watcher->addr));
 
     uv_getaddrinfo_t request;
     int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, host_end+1,
@@ -40,6 +40,7 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher, const char *endpoin
                                 });
     if (retval != 0) {
       // Failed to look up address.
+      *host_end = ':';
       goto do_pipe;
     }
     watcher->uv.tcp.addrinfo = request.addrinfo;
@@ -48,7 +49,6 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher, const char *endpoin
     watcher->stream = (uv_stream_t *)&watcher->uv.tcp.handle;
   } else {
 do_pipe:
-    xstrlcpy(watcher->addr, endpoint, sizeof(watcher->addr));
     uv_pipe_init(&loop->uv, &watcher->uv.pipe.handle, 0);
     watcher->stream = (uv_stream_t *)&watcher->uv.pipe.handle;
   }
@@ -57,8 +57,6 @@ do_pipe:
   watcher->cb = NULL;
   watcher->close_cb = NULL;
   watcher->events = NULL;
-
-  xfree(addr);
 }
 
 int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
```

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -81,33 +57,58 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+      }
+      result = uv_listen(watcher->stream, backlog, connection_cb);
+      if (result == 0) {
+        struct sockaddr_storage sas = *(struct sockaddr_storage *)ai->ai_addr;
```
Should be `struct sockaddr_storage sas;`.

mhinz replied:
Oh. Ooooh. Sorry, now I get what you meant earlier. I'll fix that.

oni-link commented:
With `:echo serverstart("127.0.0.1:12345678901")` we have pipes in `serverlist()`
that kind of look like IP addresses. Is that okay?

With `:echo serverstart("::1:1234567")` we have tcp addresses in `serverlist()` that have
a different socket address (`::1:54919`). Probably a result of how `getaddrinfo()` transforms our port string into a number (1234567 % 2^16 = 54919).

mhinz commented:
> With :echo serverstart("::1:1234567") we have tcp addresses in serverlist() that have
a different socket address (::1:54919). Probably a result of how getaddrinfo() transforms our port string into a number (1234567 % 2^16 = 54919).

`:echo serverstart("::1:1234567")` creates a Unix socket for me, as result of falling back when `getaddrinfo()` failed.

oni-link commented:
@mhinz, it creates a tcp socket for me. `uv_getaddrinfo()` does not fail. Using `nc ::1 54919 -6` and sending some text, I get an answer (`nc ::1 1234567 -6` also works for me).

mhinz commented:
@oni-link Hopefully it works as expected now.

oni-link commented:
@mhinz, port range check works.

`host:0` and `host:` both choose a random port on my system.
With `0.0.0.0:` and `0.0.0.0:0` the socket will listen on any address and picks a random port.
An empty string seems to result in a pipe with no name. Is this useful?

mhinz commented:
You mean `' '` instead of a real empty string `''`, right? The latter throws an exception for me.

It would be weird, but at least on Unix it should be no problem. I'd blame the user in that case. :> And current master acts the same, so I'd say it's fine for now.

oni-link commented:
> You mean ' ' instead of a real empty string '', right? 

No, a real empty string also works on my system (`:echo serverstart("")`) and results in an entry `''` in the output of `serverlist()`.

mhinz commented:
@oni-link 

Interesting. On my macOS box `uv_listen()` returns -13 (`UV_EACCESS`) which propagates all the way up to [`f_serverlist()`](https://github.com/neovim/neovim/blob/d76a95824d905006c0ce8f33a703ceaf63b90f9d/src/nvim/eval.c#L14269).

`serverstart()` checks for empty endpoints now.

oni-link commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -110,8 +110,8 @@ char *server_address_new(void)
-  if (endpoint == NULL) {
-    ELOG("Attempting to start server on NULL endpoint");
+  if (endpoint == NULL || endpoint[0] == '\0') {
+    ELOG("The endpoint musn't be NULL or empty.");
```
Typo, missing `t` in `musn't`.

oni-link commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -110,8 +110,8 @@ char *server_address_new(void)
-    ELOG("Attempting to start server on NULL endpoint");
+  if (endpoint == NULL || endpoint[0] == '\0') {
+    ELOG("The endpoint musn't be NULL or empty.");
     return 1;
```
Function comment says `errno` is also set. Should it be `UV_EINVAL`?

mhinz replied:
> 0 on success, 1 on a regular error, and negative errno on failure to bind or listen.

I'd argue that it should be treated as regular error, since it's input validation that happens before anything else. It's no "network error".

oni-link commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -110,8 +110,8 @@ char *server_address_new(void)
 ///          on failure to bind or connect.
 int server_start(const char *endpoint)
```
Function comment mentioning a default port could be removed.

justinmk commented:
There's no test for the changes.

mhinz committed:
Server tests: use helpers.command()

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -16,63 +16,47 @@
+    *host_end = '\0';
+    uv_getaddrinfo_t request;
+
+    int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, host_end+1,
```
Linter does not say anything about `host_end+1`. Don't we require spaces around operators (`+`)?

ZyX-I replied:
There are some exceptions to allow making some things more readable. Do not remember which exactly, AFAIR I did not touch this part much.

mhinz replied:
Personally I prefer no spaces since it makes it easier to distinguish the arguments from each other. I use spaces around operators when doing arithmetic in assignments, though.

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -81,33 +65,58 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+          : ((struct sockaddr_in6 *)&sas)->sin6_port;
+        // v:servername uses the string from watcher->addr
+        size_t len = strlen(watcher->addr);
+        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, ":%" PRIu16,
```
Spaces around `+` and `-` operators?

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -81,33 +65,58 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+        // contain 0 in this case, unless uv_tcp_getsockname() is used first.
+        uv_tcp_getsockname(&watcher->uv.tcp.handle, (struct sockaddr *)&sas,
+                           &(int){ sizeof(sas) });
+        uint16_t port = (ai->ai_family == AF_INET)
```
`ai->ai_family` could be replaced with `sas.ss_family`.

mhinz committed:
Server tests: endpoint parsing in serverstart()

mhinz commented:
@justinmk What kind of tests do you want to see?

Since there's no easy way to check what kind of sockets got created, I only added a single test that checks the endpoint parsing from `serverstart()`.

oni-link commented:
Memory leaks with `nvim +'call serverstart("::1")' +'call serverstart("::1:")' +qa`:
```c
==993==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 76 byte(s) in 1 object(s) allocated from:
    #0 0x506fb8 in malloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:66
    #1 0x31020d309c  (/lib64/libc.so.6+0x31020d309c)
    #2 0x31020d4a90 in getaddrinfo (/lib64/libc.so.6+0x31020d4a90)
    #3 0x4c5969 in __interceptor_getaddrinfo /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:2313
    #4 0x19251e8 in uv__getaddrinfo_work /home/oni-link/git/neovim/.deps/build/src/libuv/src/unix/getaddrinfo.c:102

Direct leak of 23 byte(s) in 1 object(s) allocated from:
    #0 0x506fb8 in malloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:66
    #1 0x191da5a in uv__malloc /home/oni-link/git/neovim/.deps/build/src/libuv/src/uv-common.c:75
    #2 0x191da5a in uv__strdup /home/oni-link/git/neovim/.deps/build/src/libuv/src/uv-common.c:56
    #3 0x65206d6574737972  (<unknown module>)

SUMMARY: AddressSanitizer: 99 byte(s) leaked in 2 allocation(s).
```
Problem 1: `uv_strerror()` leaks if the error is unknown (>0?):

> const char* uv_strerror(int err)
    Returns the error message for the given error code. Leaks a few bytes of memory when you call it with an unknown error code.

Problem 2: Before `server_start()` is called, two sockets can have the same name, because the port is not yet added/known.

Possible fixes:
```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index 56a6632fa..96190013b 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -14266,7 +14266,8 @@ static void f_serverstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 
   int result = server_start((char *) rettv->vval.v_string);
   if (result != 0) {
-    EMSG2("Failed to start server: %s", uv_strerror(result));
+    EMSG2("Failed to start server: %s",
+          result > 0 ? "Unknonwn system error" : uv_strerror(result));
   }
 }
 
diff --git a/src/nvim/event/socket.c b/src/nvim/event/socket.c
index 3523e8df7..2cd3c2130 100644
--- a/src/nvim/event/socket.c
+++ b/src/nvim/event/socket.c
@@ -46,11 +46,12 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
                                   .ai_family = AF_UNSPEC,
                                   .ai_socktype = SOCK_STREAM,
                                 });
+    *host_end = ':';
     if (retval != 0) {
       // Failed to look up address.
-      *host_end = ':';
       goto do_pipe;
     }
+    *(host_end + 1) = '\0';
     watcher->uv.tcp.addrinfo = request.addrinfo;
 
     uv_tcp_init(&loop->uv, &watcher->uv.tcp.handle);
@@ -96,7 +97,7 @@ int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
           : ((struct sockaddr_in6 *)&sas)->sin6_port;
         // v:servername uses the string from watcher->addr
         size_t len = strlen(watcher->addr);
-        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, ":%" PRIu16,
+        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, "%" PRIu16,
                  ntohs(port));
         break;
       }
diff --git a/src/nvim/msgpack_rpc/server.c b/src/nvim/msgpack_rpc/server.c
index 42314767f..efe63a9ea 100644
--- a/src/nvim/msgpack_rpc/server.c
+++ b/src/nvim/msgpack_rpc/server.c
@@ -124,6 +124,9 @@ int server_start(const char *endpoint)
   for (int i = 0; i < watchers.ga_len; i++) {
     if (!strcmp(watcher->addr, ((SocketWatcher **)watchers.ga_data)[i]->addr)) {
       ELOG("Already listening on %s", watcher->addr);
+      if (watcher->stream->type == UV_TCP) {
+        freeaddrinfo(watcher->uv.tcp.addrinfo);
+      }
       socket_watcher_close(watcher, free_server);
       return 1;
     }
```

oni-link commented on src/nvim/eval.c:
```diff
@@ -14266,7 +14266,8 @@ static void f_serverstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   if (result != 0) {
-    EMSG2("Failed to start server: %s", uv_strerror(result));
+    EMSG2("Failed to start server: %s",
+          result > 0 ? "Unknonwn system error" : uv_strerror(result));
```
Hm, `result` is here always `1`, so we could use the string "Unknown system error" unconditionally.

For a more specific error message we only have `errno`, but this variable is not always set on error in `server_start()`.

mhinz replied:
> `result` is here always `1`

`:call serverstart('localhost:12345')` twice throws the exception: `Failed to start server: invalid argument` (`UV_EINVAL`) for me.

oni-link replied:
Oh, read the function comment wrong that some infos are returned in `errno`. My comment above is wrong.

mhinz commented:
According to chat, there should be _no_ fallback to Unix sockets/named pipes in case a hostname lookup fails. Code needs to be adjusted.

mhinz committed:
Server: don't fall back to Unix sockets

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -40,17 +40,18 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
-    *host_end = '\0';
     uv_getaddrinfo_t request;
+    *host_end = '\0';
+    char *host = !strcmp(addr, "localhost") ? "127.0.0.1" : addr;
```
What if `localhost` is defined and its IP address is not `127.0.0.1` (perhaps a system with only IPv6 addresses that  maps `localhost` to `::1`? 
Also the the name in `serverlist()` would not match the address that the socket is using. Should names in the list be replaced with the IP address that is used with the tcp socket?

mhinz replied:
What about this:

```diff
diff --git i/src/nvim/event/socket.c w/src/nvim/event/socket.c
index d07b81bc2..895dfdca0 100644
--- i/src/nvim/event/socket.c
+++ w/src/nvim/event/socket.c
@@ -42,17 +42,25 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,

     uv_getaddrinfo_t request;
     *host_end = '\0';
-    char *host = !strcmp(addr, "localhost") ? "127.0.0.1" : addr;

-    int retval = uv_getaddrinfo(&loop->uv, &request, NULL, host, port,
+    char host[sizeof(watcher->addr)];
+    xstrlcpy(host, addr, sizeof(host));
+
+getaddrinfo:
+    ret = uv_getaddrinfo(&loop->uv, &request, NULL, host, port,
                                 &(struct addrinfo){
                                   .ai_family = AF_UNSPEC,
                                   .ai_socktype = SOCK_STREAM,
                                 });
     *host_end = ':';
-    if (retval != 0) {
+    if (ret != 0) {
+      if (strcmp(host, "localhost") == 0) {
+        xstrlcpy(host, "127.0.0.1", sizeof(host));
+        snprintf(watcher->addr, sizeof(watcher->addr), "%s:%s", host, port);
+        goto getaddrinfo;
+      }
       ELOG("Host lookup failed: %s", host);
-      return retval;
+      return ret;
     }
     *port = '\0';
     watcher->uv.tcp.addrinfo = request.addrinfo;
```

oni-link replied:
If we use that `localhost` and `127.0.0.1` have the same length, the new array `host` is not needed:
```diff
diff --git a/src/nvim/event/socket.c b/src/nvim/event/socket.c
index d07b81bc2..bc80adc53 100644
--- a/src/nvim/event/socket.c
+++ b/src/nvim/event/socket.c
@@ -40,20 +40,24 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
       return UV_EINVAL;
     }
 
-    uv_getaddrinfo_t request;
     *host_end = '\0';
-    char *host = !strcmp(addr, "localhost") ? "127.0.0.1" : addr;
+    uv_getaddrinfo_t request;
 
-    int retval = uv_getaddrinfo(&loop->uv, &request, NULL, host, port,
+getaddrinfo:
+    ret = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,
                                 &(struct addrinfo){
                                   .ai_family = AF_UNSPEC,
                                   .ai_socktype = SOCK_STREAM,
                                 });
-    *host_end = ':';
-    if (retval != 0) {
-      ELOG("Host lookup failed: %s", host);
-      return retval;
+    if (ret != 0) {
+      ELOG("Host lookup failed: %s", addr);
+      if (strcmp(addr, "localhost") == 0) {
+        xstrlcpy(addr, "127.0.0.1", sizeof(addr));
+        goto getaddrinfo;
+      }
+      return ret;
     }
+    *host_end = ':';
     *port = '\0';
     watcher->uv.tcp.addrinfo = request.addrinfo;
 

```

jamessan replied:
Why are we jumping through all these hoops for the theoretical case of someone specifying `localhost` when that isn't a known hostname on their system?  What's the goal here?

mhinz replied:
Convenience, I hoped. But I really don't care. Should I leave away this commit?

I'd do anything to get this PR merged and end the back-and-forth. ;]

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -43,16 +43,21 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
       ELOG("Host lookup failed: %s", addr);
-      return retval;
+      if (strcmp(addr, "localhost") == 0) {
+        xstrlcpy(addr, "127.0.0.1", sizeof(addr));
```
`sizeof(addr)` is the pointer size, not the size of the array `addr` points to.

jamessan commented on src/nvim/event/socket.c:
```diff
@@ -43,16 +43,21 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
       ELOG("Host lookup failed: %s", addr);
-      return retval;
+      if (strcmp(addr, "localhost") == 0) {
+        xstrlcpy(addr, "127.0.0.1", sizeof(*addr));
```
`sizeof(*addr)` is 1, the size of a `char`.  You would need `strlen() + 1`, but I think we should just remove the special casing for localhost.  If a user can't specify a valid hostname, then they should get an error.

mhinz replied:
Okay, I dropped the commit.

mhinz commented:
1. **Unix**: Everything works and the tests pass.
1. **Windows**: Even a simple `serverstart('127.0.0.1:12345')` fails [according to CI](https://ci.appveyor.com/project/neovim/neovim/build/13469/job/1o0ds2gjs7cald9l#L4552). I'm grateful for any help here, since I have no idea about Windows.

jamessan commented on src/nvim/event/socket.c:
```diff
@@ -81,33 +66,58 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+        break;
+      }
+    }
+    freeaddrinfo(watcher->uv.tcp.addrinfo);
```
This must be `uv_freeaddrinfo`, not `freeaddrinfo`.

jamessan commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -97,31 +97,36 @@ char *server_address_new(void)
     if (!strcmp(watcher->addr, ((SocketWatcher **)watchers.ga_data)[i]->addr)) {
       ELOG("Already listening on %s", watcher->addr);
+      if (watcher->stream->type == UV_TCP) {
+        freeaddrinfo(watcher->uv.tcp.addrinfo);
```
This must be `uv_freeaddrinfo`, not `freeaddrinfo`.

jamessan reviewed (commented):
The changes mentioned below fix the crashes on Windows.  I'd also suggest changing the wording of 013268e7 to state `uv_getaddrinfo` instead of `getaddrinfo`.

mhinz commented:
@oni-link @bfredl Can this be merged?

justinmk commented on src/nvim/event/socket.c:
```diff
@@ -23,42 +23,43 @@
+
+    int ret = getdigits_safe(&(char_u *){ (char_u *)port }, &iport);
+    if (ret == FAIL || iport < 0 || iport > UINT16_MAX) {
+      ELOG("Invalid port: %s", port);
```
missing log.h include?

mhinz replied:
Seemed to work without it. Not sure which of the headers include it already.

jamessan replied:
Explicit is better than implicit. :-)

mhinz replied:
Ok, added.

justinmk commented on src/nvim/event/socket.c:
```diff
@@ -17,60 +17,46 @@
+                                  .ai_family = AF_UNSPEC,
+                                  .ai_socktype = SOCK_STREAM,
+                                });
+    *host_end = ':';
```
what's the purpose of this and the assignment on line 44?

mhinz replied:
Even if it makes the code longer? Because we need to keep the `host_end` in any case:

```c
char *host_end = strrchr(addr, ':');
if (host_end) {
  ptrdiff_t sepoff = host_end - addr;
  ...
  addr[sepoff] = '\0'
```

Something like that? Is that better?

justinmk replied:
It's only 1 AFAICT.

    ptrdiff_t off = host_end ? (host_end - addr) : 0;

mhinz replied:
But then `addr[addr_off] = '\0'` would be wrong in case of `addr_off == 0`?

justinmk replied:
Why? Where would `host_end` be pointing to in that case? 

Also still wondering about line ~~44~~ 45.

mhinz replied:
`uv_getaddrinfo()` wants two strings for `host` and `port`. We have `addr == `host:port`. `host_end` points to the address of `:` in that string.

Basically we make 1 string 2 strings for some time to avoid allocating space for a new string.

justinmk replied:
Understood, but that doesn't answer my questions.

mhinz replied:
For `uv_getaddrinfo()`, we want `addr` to have the format `"host"`, and for `ELOG()`, we want `addr` to have the format `"host:port"`.

If that doesn't answer it, then I don't understand the question.

justinmk replied:
Assuming this optimization is worth it, the extra code for the offset approach is worth it. That at least reduces the cleverness by 1 unit :)

oni-link replied:
How about we add a comment and remove `*host_end=':';` and `*port='\0';`? This code is not needed anymore, because we don't continue in the "pipe" part.
```diff
diff --git a/src/nvim/event/socket.c b/src/nvim/event/socket.c
index 8509bb1ea..3a47f3489 100644
--- a/src/nvim/event/socket.c
+++ b/src/nvim/event/socket.c
@@ -33,6 +33,9 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
   char *host_end = strrchr(addr, ':');
 
   if (host_end && addr != host_end) {
+    // Split user specified address into two strings, addr(hostname) and port.
+    // The port part in watcher->addr will be updated later.
+    *host_end = '\0';
     char *port = host_end + 1;
     intmax_t iport;
 
@@ -42,7 +45,6 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
       return UV_EINVAL;
     }
 
-    *host_end = '\0';
     uv_getaddrinfo_t request;
 
     int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,
@@ -50,12 +52,10 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
                                   .ai_family = AF_UNSPEC,
                                   .ai_socktype = SOCK_STREAM,
                                 });
-    *host_end = ':';
     if (retval != 0) {
       ELOG("Host lookup failed: %s", addr);
       return retval;
     }
-    *port = '\0';
     watcher->uv.tcp.addrinfo = request.addrinfo;
 
     uv_tcp_init(&loop->uv, &watcher->uv.tcp.handle);
@@ -102,7 +102,7 @@ int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
           : ((struct sockaddr_in6 *)&sas)->sin6_port;
         // v:servername uses the string from watcher->addr
         size_t len = strlen(watcher->addr);
-        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, "%" PRIu16,
+        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, ":%" PRIu16,
                  ntohs(port));
         break;
       }

```

mhinz replied:
@oni-link Hmm, right, and then just using `endpoint` in the `ELOG()` to keep the `host:port` format.

mhinz committed:
Doc: explain the format for serverstart()

justinmk commented on src/nvim/event/socket.c:
```diff
@@ -82,33 +68,60 @@ void socket_watcher_init(Loop *loop, SocketWatcher *watcher,
+        // v:servername uses the string from watcher->addr
+        size_t len = strlen(watcher->addr);
+        snprintf(watcher->addr+len, sizeof(watcher->addr)-len, "%" PRIu16,
+                 ntohs(port));
```
why would the byte order of `sin_port` be different than the host system?

jamessan replied:
Because it's always network order.

mhinz replied:
The network byte order is big endian and my host's byte order is little endian.

justinmk commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -97,37 +97,47 @@ char *server_address_new(void)
-  if (endpoint == NULL) {
-    ELOG("Attempting to start server on NULL endpoint");
+  if (endpoint == NULL || endpoint[0] == '\0') {
+    ELOG("The endpoint mustn't be NULL or empty.");
```
typo. Also avoid "The", it's usually useless at the start of a message or comment.

    ELOG("Empty or NULL endpoint");

mhinz replied:
The used to be fixed in a later commit that was dropped..

justinmk commented on src/nvim/msgpack_rpc/server.c:
```diff
@@ -97,37 +97,47 @@ char *server_address_new(void)
+  result = socket_watcher_start(watcher, MAX_CONNECTIONS, connection_cb);
   if (result < 0) {
     ELOG("Failed to start server: %s", uv_strerror(result));
     socket_watcher_close(watcher, free_server);
```
is `uv_freeaddrinfo(watcher->uv.tcp.addrinfo);` needed here also? Or should that be done in `socket_watcher_close` ?

mhinz replied:
When `socket_watcher_start()` returns something `< 0`, then `uv_getaddrinfo()` failed or an invalid port was specificed. No need for `uv_freeaddrinfo()` in that case.

jamessan commented:
The Linux QB failure is

```
11:42:41,907 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/event/socket.c: In function ‘socket_watcher_start’:
11:42:41,907 WARN  - /home/quickbuild/buildagent/workspace/root/neovim/pull-requests-automated/src/nvim/event/socket.c:102:11: error: conversion to ‘uint16_t’ from ‘int’ may alter its value [-Werror=conversion]
```

which doesn't make sense because both `sin6_port` and `sin_port` should be the same type -- `in_port_t`.  Everywhere I've check, that _is_ a typedef to `uint16_t`.  However, this is happening on Ubuntu 12.04, which is EOL, so I can't see what version of glibc it was using.

The FreeBSD QB failure is

```
    funcs.serverstart('127.0.0.1:')  -- assign random port
    assert(string.match(funcs.serverlist()[1], '127.0.0.1:%d+'))
```

on the assert:

```
19:17:33,940 INFO  - # test/functional/eval/server_spec.lua @ 61
19:17:33,940 INFO  - # Failure message: test/functional/eval/server_spec.lua:70: bad argument #1 to 'match' (string expected, got nil)
19:17:33,940 INFO  - # stack traceback:
19:17:33,940 INFO  - #     test/functional/eval/server_spec.lua:70: in function <test/functional/eval/server_spec.lua:61>
```

It'd be interesting to see what `serverlist()` is actually returning.

bfredl commented on runtime/doc/eval.txt:
```diff
@@ -6462,11 +6462,20 @@ serverlist()						*serverlist()*
+		everything after it the port. If the port is empty or `0`,
+		a random port will be assigned.
+
+		If no address is given, it is equivalent to: >
```
is this true also on windows (pipes have a disjoint namespace from regular files)?

mhinz replied:
I have no idea. I trust the person who wrote it in the first place.

bfredl replied:
I wouldn't be so trusting :P

oni-link replied:
For Windows the special name format `\\.\pipe\nvim-<PID>-<COUNTER>` is used.

bfredl commented on src/nvim/event/socket.c:
```diff
@@ -17,60 +17,47 @@
+    uv_getaddrinfo_t request;
+
+    int retval = uv_getaddrinfo(&loop->uv, &request, NULL, addr, port,
+                                &(struct addrinfo){
```
why not declare hints as a separate var?

mhinz replied:
Well, there's no need to keep it, so I used a throwaway compound literal.

mhinz commented:
> It'd be interesting to see what serverlist() is actually returning.
 
Hmm, if `serverlist()[1]` is `nil`, then `serverlist()` probably returned `{}`, right?

So `serverstart('127.0.0.1:')` failed whereas `serverstart('127.0.0.1:0')` succeeded?

bfredl commented:
Wouldn't it be useful if `serverstart("localhost:")` returned the actual used address, i e `"localhost:46565"` ?

mhinz commented:
@bfredl Absolutely. In a follow-up PR. :-)

@jamessan In that case I'll just ignore the Linux QB issue.

What to do about the FreeBSD QB issue? Should I just put logging into the test until we can figure it out? Is anyone using FreeBSD by any chance?

bfredl commented:
Why? You already calculate the correct value for serverlist, why not return it already?

jamessan committed:
Server: Call uv_getaddrinfo with NULL service when no port

When using serverstart("ip.ad.d.r:") to listen on a random port, we need
to abide by getaddrinfo()'s API and pass in a NULL service, rather than
an empty string.

When given an empty string, getaddrinfo() is free to search for a
service by the given name (since the string isn't a number) which will
fail.  At least FreeBSD does perform this lookup.

jamessan commented:
Hmm, while testing the FreeBSD fix I ran into this:

```
==5748==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 2 byte(s) in 1 object(s) allocated from:
    #0 0x4ecdc8 in malloc (/home/jamessan/src/github.com/neovim/build/bin/nvim+0x4ecdc8)
    #1 0xf828e4 in try_malloc /home/jamessan/src/github.com/neovim/src/nvim/memory.c:87:15
    #2 0xf82aa4 in xmalloc /home/jamessan/src/github.com/neovim/src/nvim/memory.c:121:15
    #3 0xf82f43 in xmallocz /home/jamessan/src/github.com/neovim/src/nvim/memory.c:196:15
    #4 0xf83038 in xmemdupz /home/jamessan/src/github.com/neovim/src/nvim/memory.c:214:17
    #5 0xf83e92 in xstrdup /home/jamessan/src/github.com/neovim/src/nvim/memory.c:429:10
    #6 0x175865d in vim_strsave /home/jamessan/src/github.com/neovim/src/nvim/strings.c:59:20
    #7 0xb2bdd9 in do_cmdline /home/jamessan/src/github.com/neovim/src/nvim/ex_docmd.c:553:22
    #8 0xb32b65 in do_cmdline_cmd /home/jamessan/src/github.com/neovim/src/nvim/ex_docmd.c:279:10
    #9 0xe6f0f1 in exe_pre_commands /home/jamessan/src/github.com/neovim/src/nvim/main.c:1607:7
    #10 0xe60ee8 in main /home/jamessan/src/github.com/neovim/src/nvim/main.c:335:3
    #11 0x7f74cce1a2b0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202b0)

SUMMARY: AddressSanitizer: 2 byte(s) leaked in 1 allocation(s).
+++ exited with 1 +++
```

by running `./build/bin/nvim -u NONE -i NONE -n --cmd 'call serverstart("127.0.0.1:")' --cmd q`.  I tried a few other two "pre_command" command lines, but none of those leaked.  I can reproduce it on master using `serverstart()`...

jamessan committed:
eval: serverstart: Return finalized address to user

In the process of setting up the socket watcher, the address may be
changed (e.g., adding the OS-selected port).

jamessan committed:
socket_watcher_start: Silence conversion warning for sin(6)_port

Although in_port_t is a typedef for uint16_t, GCC in Ubuntu 12.04
complains about potential loss of data due to converting int to
uint16_t.  Since we know this isn't possible, silence the warning to
avoid breaking QB until it gets upgraded to a newer Ubuntu.

jamessan commented:
> In that case I'll just ignore the Linux QB issue.

Unfortunately, that would mean QB is red for everyone until we get off of that Ubuntu release.  @jszakmeister is there any chance the QB Linux systems could be upgraded to at least Ubuntu 14.04 (Trusty)?

In the meantime, I've added a cast which should silence this.

> What to do about the FreeBSD QB issue?

I've pushed a fix for that.

> Wouldn't it be useful if `serverstart("localhost:")` returned the actual used address, i e `"localhost:46565"`?

Agreed.  I've also pushed a commit for this.

oni-link commented on src/nvim/event/socket.c:
```diff
@@ -45,6 +45,12 @@ int socket_watcher_init(Loop *loop, SocketWatcher *watcher,
       return UV_EINVAL;
     }
 
+    if (*port == NUL) {
```
Should this also be used if `iport` is `0` or is `"0"` always an acceptable port description?

jamessan replied:
A port that looks like a number is used directly, otherwise a lookup is performed.  We should be fine here.

oni-link commented on src/nvim/eval.c:
```diff
@@ -14296,23 +14296,41 @@ static void f_serverstart(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+  // "localhost:" will now have a port), return the final value to the user.
+  size_t n;
+  char **addrs = server_address_list(&n);
+  address = addrs[n - 1];
```
Why use `address` instead of `rettv->vval.v_string`?

jamessan replied:
Fixed.

jszakmeister commented:
> @jszakmeister is there any chance the QB Linux systems could be upgraded to at least Ubuntu 14.04 (Trusty)?

@jamessan Yes I can.  The real question, which has never been answered though, is what is our support policy?  And I guess one of the additional headaches here is "what do we want to ensure we run against" vs. "what do we want to build against with no warnings"?

Let me know what you guys want here (and it's generally best to do this via email directly to me--notifications from GitHub have a tendency to get lost in the noise).
